File: code/copy_list_with_random_pointer.java
/**
 * Definition for singly-linked list with a random pointer.
 * class RandomListNode {
 *     int label;
 *     RandomListNode next, random;
 *     RandomListNode(int x) { this.label = x; }
 * };
 */
public class Solution {
    /**
     * @param head: The head of linked list with a random pointer.
     * @return: A new head of a deep copy of the list.
     */
    public RandomListNode copyRandomList(RandomListNode head) {
        // write your code here

        RandomListNode dummy = new RandomListNode(0);

        RandomListNode origin = head;
        RandomListNode prev = dummy;
 
        HashMap<RandomListNode, RandomListNode> map = new HashMap<>();

        while (origin != null) {
            RandomListNode node = new RandomListNode(origin.val);
            map.put(origin, node);
            prev.next = node;
            prev = prev.next;
            origin = origin.next;
        }

        origin = head;
        while (origin != null) {
            map.get(origin).random = map.get(origin.random);
        }

        return dummy.next;

    }
}

File: code/medium1/copy_list_with_random_pointer.cpp
#include <iostream>
#include <unordered_map>
using namespace std;
/**
 * Definition for singly-linked list with a random pointer.
 * struct RandomListNode {
 *     int label;
 *     RandomListNode *next, *random;
 *     RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
 * };
 */

struct RandomListNode {
      int label;
      RandomListNode *next, *random;
      RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
};

class Solution {
public:
    RandomListNode *copyRandomList(RandomListNode *head) {

        unordered_map<RandomListNode*, RandomListNode*> nodeMap;
        RandomListNode* anchor = head;
        while (head != NULL) {
            RandomListNode* node = new RandomListNode(head->label);
            nodeMap[head] = node;
            head = head->next;
        }
        head = anchor;
        while (head != NULL) {
            RandomListNode* curr = nodeMap[head];
            curr->random = nodeMap[head->random];
            head = head->next;
        }

        return nodeMap[anchor];
    }
};